conn.go

```diff
@@ -2,52 +2,127 @@ package conn
 
 import (
 	"context"
+	"errors"
 	"io"
 	"net"
-	"time"
 
 	logging "github.com/ipfs/go-log"
-	mpool "github.com/jbenet/go-msgio/mpool"
+	ci "github.com/libp2p/go-libp2p-crypto"
 	ic "github.com/libp2p/go-libp2p-crypto"
 	iconn "github.com/libp2p/go-libp2p-interface-conn"
 	lgbl "github.com/libp2p/go-libp2p-loggables"
 	peer "github.com/libp2p/go-libp2p-peer"
+	secio "github.com/libp2p/go-libp2p-secio"
 	tpt "github.com/libp2p/go-libp2p-transport"
+	smux "github.com/libp2p/go-stream-muxer"
 	ma "github.com/multiformats/go-multiaddr"
 )
 
 var log = logging.Logger("conn")
 
-// ReleaseBuffer puts the given byte array back into the buffer pool,
-// first verifying that it is the correct size
-func ReleaseBuffer(b []byte) {
-	log.Debugf("Releasing buffer! (cap,size = %d, %d)", cap(b), len(b))
-	mpool.ByteSlicePool.Put(uint32(cap(b)), b)
-}
-
-// singleConn represents a single connection to another Peer (IPFS Node).
+// singleConn represents a single stream-multipexed connection to another Peer (IPFS Node).
 type singleConn struct {
-	local  peer.ID
-	remote peer.ID
-	maconn tpt.Conn
-	event  io.Closer
-}
+	streamConn smux.Conn
+	tptConn    tpt.Conn
+
+	secSession secio.Session
+
+	event io.Closer
+}
+
+var _ iconn.Conn = &singleConn{}
+
+// newSingleConn constructs a new connection
+func newSingleConn(ctx context.Context, local, remote peer.ID, privKey ci.PrivKey, tptConn tpt.Conn, pstpt smux.Transport, isServer bool) (iconn.Conn, error) {
+	ml := lgbl.Dial("conn", local, remote, tptConn.LocalMultiaddr(), tptConn.RemoteMultiaddr())
+
+	var streamConn smux.Conn
+	var secSession secio.Session
+	switch conn := tptConn.(type) {
+	case tpt.DuplexConn:
+		c := conn
+		// 1. secure the connection
+		if privKey != nil && iconn.EncryptConnections {
+			var err error
+			secSession, err = setupSecureSession(ctx, local, privKey, conn)
+			if err != nil {
+				return nil, err
+			}
+			c = &secureConn{
+				DuplexConn: conn,
+				secure:     secSession,
+			}
+		} else {
+			log.Warning("creating INSECURE connection %s at %s", tptConn.LocalMultiaddr(), tptConn.RemoteMultiaddr())
+		}
 
-// newConn constructs a new connection
-func newSingleConn(ctx context.Context, local, remote peer.ID, maconn tpt.Conn) (iconn.Conn, error) {
-	ml := lgbl.Dial("conn", local, remote, maconn.LocalMultiaddr(), maconn.RemoteMultiaddr())
+		// 2. start stream multipling
+		var err error
+		streamConn, err = pstpt.NewConn(c, isServer)
+		if err != nil {
+			return nil, err
+		}
+	case tpt.MultiplexConn:
+		// 1. secure the connection on the first best stream
```

Comment:
We can also do as you suggest and treat each stream as a separate connection and protect each one.

Regardless, whatever we do, this has implications on the transport interfaces. IMO, we should require that *all* multiplexed connections provide their own security (which means we may need to change that interface...).

Reply:
Actually, if we *do* assume that all multiplexed transports are secure, we should probably push the security stuff into the quick smux library.

Reply:
I like your last idea. I think it's a reasonable assumption that all multiplexed transports have their own security mechanism (QUIC is the first multiplexed transport we have, and all future ones will only get better than that).

I started working on the TLS 1.3 integration into QUIC last week, and I have a very basic version working now (still far from production use though). Once I'm done with this, this will allow us to get rid of secio (and probably also the protector, if we want to).

We still have to decide what we do here. I'm not sure what a good solution would look like. Maybe we should just panic here.

Reply:
For now, panicing sounds reasonable; we don't want users using this at this point. We'll need to fix the MultiplexConn interface regardless of what we do (to add security related methods) so it isn't really usable as-is anyways.

---

After thinking about all this, I think there may be a better way to structure this (but we can do this later unless you think this will be quick and easy). Basically, we can make *all* transports return multiplexed, secured connections and then provide convenience methods to provide security/multiplexing for protocols that don't have built-in support. For example, the TCP transport would be defined as:

```go
type tcpDialer struct {/* ... */}

func (t *tcpDialer) Dial(raddr ma.Multiaddr) {
    c, err := tcp.Dial(...)
    if err != nil {
        return err
    }
    sc, err := secio.Secure(c)
    if err != nil {
        return err
    }
    return smux.Multiplex(sc) // negotiates the multiplexer.
}
```

IMO, this better fits go's composition over inheritance pattern.